### ISSUE TITLE
`EventID` incorrectly inherits from `DomainSpecificString`

### ISSUE THREAD

**Issue by clokep** [matrix-org/synapse#14638]
> Not super happy with the `startswith` but `EventID.is_valid` is broken for non-V1/2 event IDs so unsure how to proceed.

_Originally posted by @Fizzadar in https://github.com/matrix-org/synapse/pull/14632#discussion_r1042247889_

Modern event IDs are now hashes and do not include a domain name, so they shouldn't inherit from `DomainSepcificString`.

**Comment by clokep** [comment]
> good spot. We should define a new string type for this with a corrected grammar?

Yep! I don't know if there's anything that would break if it no longer has the same interface as `DomainSpecificString`.

I think the grammar is just `$[...characters...]` unfortunately.

**Comment by reivilibre** [comment]
That wouldn't surprise me, I guess. Probably null bytes might be accidentally excluded because of database support but everything else might be acceptable.
v3+ event IDs have a more strict grammar owing to the fact they are an encoded hash, so we could define an alphabet for those. For the non-v3+ event IDs, we can enforce that there is a domain.

**Comment by DMRobertson** [comment]
> v3+ event IDs have a more strict grammar owing to the fact they are an encoded hash, so we could define an alphabet for those. 

I think they also have an expected length?

>  For the non-v3+ event IDs, we can enforce that there is a domain.

Similarly, I think the spec might define a maximum length of an event ID here
